add core.sync.event to be used as Windows events

rainers opened the pull request:
extracted from https://github.com/dlang/druntime/pull/2514

jacob-carlborg commented on src/core/sync/event.d:
```diff
@@ -0,0 +1,318 @@
+
+version (Windows)
+{
+    import core.sys.windows.basetsd /+: HANDLE +/;
```
Why these comments?

rainers replied:
This follows the example of https://github.com/dlang/druntime/pull/2400/files (selective imports create a public alias).

wilzbach replied:
Hasn't this been fixed a year ago or so?

wilzbach replied:
Oh well, looks like it's really still in deprecation, but there's a PR to remove them for good:

https://github.com/dlang/dmd/pull/9393

thewilsonator replied:
That will go in very soon.

jacob-carlborg replied:
@rainers this has been merged now.

rainers replied:
Not good  enough:
```
test.d(3): Deprecation: core.sync.event.HANDLE is not visible from module test
```

jacob-carlborg replied:
😞

jacob-carlborg commented on src/core/sync/event.d:
```diff
@@ -0,0 +1,318 @@
+                pthread_cond_signal(&waiter.cond);
+                pthread_mutex_unlock(&waiter.mutex);
+            }
+            pthread_mutex_unlock(&m_mutex);
```
`scope(exit)`?

rainers replied:
The function is nothrow anyway, so it doesn't make a difference in  code generation. With a short function like this, having the symmetry show up in the function is slightly better IMO.

jacob-carlborg commented on src/core/sync/event.d:
```diff
@@ -0,0 +1,318 @@
+
+    /**
+     * Set the event to "signaled", so that waiting clients are resumed
+     */
```
Using `///` would save two lines.

rainers replied:
Ok.

jacob-carlborg commented on src/core/sync/event.d:
```diff
@@ -0,0 +1,318 @@
+/**
+ * The event module provides a primitive for lightweight signalling of other threads
+ * (aka Windows Events)
```
Reading this, I though the implementation was for Windows only.

rainers replied:
reworded slightly.

jacob-carlborg commented on src/core/sync/event.d:
```diff
@@ -0,0 +1,318 @@
+import core.internal.abort : abort;
+
+/**
+ * This class represents an event. Clients of an event are suspended while waiting
```
This is a struct.

rainers replied:
Ok. It started out as a class, but no longer.

jacob-carlborg commented on src/core/sync/event.d:
```diff
@@ -0,0 +1,318 @@
+     *  manualReset  = the state of the event is not reset automatically after resuming waiting clients
+     *  initialState = initial state of the signal
+     */
+    void initialize(bool manualReset, bool initialState)
```
Why is this a separate method?

rainers replied:
You cannot pass arguments to the constructor if it is part of another struct, e.g. https://github.com/dlang/druntime/pull/2514/files#diff-6615611ba51118215a271e3500c61122R2605. You have to use an explicit `initialize`-call then. `__ctor` could be an alternative, but I think it should not be part of the API.

jacob-carlborg replied:
Is there a downside to do `evStart = Event(...)` in `Gcx.initialize`?

rainers replied:
I've explicitly added `@disable this(this)` as this could duplicate the handle reference.

jacob-carlborg replied:
Does not seem to prevent assignment: https://run.dlang.io/is/iefkBG.

rainers replied:
Hmm, seems like opAssign should be disabled, too.

jacob-carlborg replied:
If that's required then I'm fine with using `initialize` instead of a constructor.

jacob-carlborg commented on src/core/sync/event.d:
```diff
@@ -0,0 +1,314 @@
+        terminate();
+    }
+
+    void terminate()
```
No documentation.

jacob-carlborg commented:
No changelog.

rainers commented:
Added changelog and a bit more documentation. I've also added a bool to keep track of the initialization state of the posix version.

>     * uncommented selective imports

I think this should be done for all the imports once the dmd deprecation is gone (https://github.com/dlang/dmd/pull/9393 doesn't help for this case yet).

thewilsonator commented:
`posix.mak:348: recipe for target 'checkwhitespace' failed`

jacob-carlborg commented on changelog/core_sync_event.dd:
```diff
@@ -0,0 +1,34 @@
+New module core.sync.event
+
+This module provides a primitive for lightweight signaling of other threads
+by emulating Windows events on posix. It can be used to start execution of
```
* Perhaps I'm picky, but now it sounds like it's for Posix only. I would write something like: "... a cross-platform interface  for lightweight signaling of other threads. "
* `posix` -> `Posix`

rainers replied:
No problem ;-) Amended.

kinke commented on changelog/core_sync_event.dd:
```diff
@@ -0,0 +1,33 @@
+New module core.sync.event
+
+This module provides a cross-platform interface for lightweight signaling of other threads.
+It can be used to start execution of multiple waiting threads simultaniously.
```
simultan**e**ously

kinke commented:
Rainer, have you explored whether [pthread_cond_broadcast](http://pubs.opengroup.org/onlinepubs/009696699/functions/pthread_cond_broadcast.html) and a single condition variable could work too?

rainers commented:
> Rainer, have you explored whether [pthread_cond_broadcast](http://pubs.opengroup.org/onlinepubs/009696699/functions/pthread_cond_broadcast.html) and a single condition variable could work too?

Didn't know about it, now replaced the manual implementation.

wilzbach commented on src/core/sync/event.d:
```diff
@@ -0,0 +1,301 @@
+    }
+}
+
+// Test single-thread (non-shared) use.
```
Maybe make this example public, s.t. there's at least one user-facing example.

rainers replied:
I think this example is a pretty bad one as it doesn't involve multiple threads. Maybe better to move the example from the changelog into the documentation?

wilzbach replied:
Sure. I just saw that the module is currently without any public example.

Geod24 commented on src/core/sync/event.d:
```diff
@@ -37,6 +37,35 @@ import core.internal.abort : abort;
+        event.terminate();
+    }
+}
+---
```
Why isn't it a documented unittest instead ?

rainers replied:
I think its already a bit too long for a really good example, and making it actually runnable would make it even longer.
